datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java

```diff
@@ -222,14 +248,22 @@ ZkClient createZkClient() {
     return new ZkClient(_zkServers, _sessionTimeoutMs, _connectionTimeoutMs, _operationRetryTimeoutMs);
   }
 
+  @VisibleForTesting
+  ZkStateChangeListener getOrCreateStateChangeListener() {
+    if (_stateChangeListener == null) {
+      _stateChangeListener = new ZkStateChangeListener();
+    }
+    return _stateChangeListener;
+  }
+
   /**
    * Connect the adapter so that it can connect and bridge events between ZooKeeper changes and
    * the actions that need to be taken with them, which are implemented in the Coordinator class
    */
   public void connect() {
     disconnect(); // Guard against leaking an existing zookeeper session
```

Comment:
I will do it in next pass. Removing it will make the situation worse, if this PR has any bugs.

Reply:
If you call disconnect without connect, it can lead to leaderless scenario till we fix everything. [previous analysis](https://docs.google.com/document/d/14rhn6U6AOqZ_swkQsuUKzH7DCS7Ab3aSvWcuXhESBEI/edit#heading=h.syjnl87mxqob) where we had zk sessions leak when disconnect was not called before connect.

Reply:
Got it. How about resetting the current thread's interrupted flag after calling `disconnect()` so the subsequent `createZkClient()` can proceed?

Reply:
Adding this change as part of this PR can open another set of issues. The behavior will not change much after the current PR. I will consider this when we are ready to enable reconnect and sure that there are no other loose ends.

Reply:
OK. So, just to confirm my understanding, this PR is not meant to handle the following scenario:
1. ZK session expires. Out code observes a `Disconnected` state and schedules an `onSessionExpired()` call
2. `onSessionExpired()` gets invoked after the timeout elapses and `onBecomeFollower()` is called
3. ZK client reestablishes connectivity and its connection state transitions to `Expired`
4. ZK client's `fireAllEvents()` is invoked, calls `joinLeaderElection()`
5. The leader election code fails to find the current instance name among the live instances, calls `connect()` which calls `disconnect()` followed by `createZkClient()` — an exception is thrown and the host gets in an undefined state.

Reply:
The sequence of the scenario is right, but we have more clarity on point 5.

1. Leader node will not run as leader node anymore (So, we will not see multiple leaders problem)
2. We will have metrics in next PR that we will use as alert to restart the node (which is not happening today)
3. The zookeeper states will be fine.

Next step will be:
1. Add metrics and alert.
2. Task cleanup support within fixed time.
3. Add support to handle low-level kafka.
4. Then enable the reconnection.

Reply:
@vmaheshw Personally, it'll be much easier to review any changes related to ZK session expiry once I've seen your design and implementation plan. It can avoid unnecessary confusion like this. It'd be nice if you can prioritize working on that before sending out further PRs. It will also help with prioritizing the next steps.